doc/TclObject.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH TclObject 3 1.0 Tcl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclObject -- the basic notion of an "object" in Tcl.

.SH CONSTRUCTORS
TclObject

.SH METHODS
.sp
\fBTclObject\fR(InternalRep \fIrep\fR)
.sp
void \fBsetInternalRep\fR(InternalRep \fIrep\fR)
.sp
InternalRep \fBgetInternalRep\fR()
.sp
String \fBtoString\fR()
.sp
void \fBinvalidateStringRep\fR()
.sp
void \fBpreserve\fR()
.sp
void \fBrelease\fR()
.sp
boolean \fBisShared\fR()
.sp
TclObject \fBtakeExclusive\fR()

.SH ARGUMENTS
.AS InternalRep rep .
.AP InternalRep rep
Internal representation to be stored in newly created TclObject.
.BE

.SH DESCRIPTION
The Java API to the Tcl interpreter is object-based -- in most cases,
the values passed to and from the Tcl interpreter are instances of
TclObject. This includes variable values and command arguments.
.PP
A TclObject is dual-ported: it behaves like a String but also holds an
internal representation that can be manipulated more efficiently.  For
example, a TclList is represented as a TclObject that holds the
list's string representation as well as a Vector to hold the
objects for each list element. Dual-ported objects avoid most runtime
type conversions.  This improves the speed of many operations
since an appropriate representation is immediately available.
.PP
Currently, the following internal representations are supported:
TclBoolean, TclDouble, TclList, TclIndex, TclInteger, TclString and
ReflectObject. Most of these internal representations have a
newInstance() method, which can be used to create a new TclObject
instance that contains the specific internal representation.  You
should always create TclObject instances with the newInstance() 
methods; use the "new" operator to create TclObject instances only
when you are writing new internal representations.
.PP
The type of the internal representation of Tcl objects can mutate.
Methods such as \fBsetInternalRep\fR and \fBinvalidateStringlRep\fR
are facilities for mutating the internal rep.
.PP
To improve memory efficiency, TclObject supports copy-on-write
operations.  When you need to save the value of a TclObject for later
use, call the \fBpreserve\fR method; when you no longer need its
value, call the \fBrelease\fR method. Internally, each call to
\fBpreserve\fR will internally increment the "reference count" of the
TclObject by one; conversely, each call to \fBrelease\fR decrements
the reference count by one.
.PP
Some methods of classes derrived from the \fBInternalRep\fR class,
such as \fBTclList\fR.\fBappend()\fR and \fBTclInteger\fR.\fBset()\fR,
modify the contents of a TclObject.  When you are about to call these
methods to modify the value of a TclObject, you must call \fBpreserve\fR (if
you haven't already done so) and then call \fBtakeExclusive\fR.
\fBtakeExclusive\fR will make a copy of the TclObject if it is shared.
The TclObject returned by the \fBtakeExclusive\fR will always have a
reference count of 1.
.TP
\fBTclObject\fR
The \fBTclObject\fR constructor creates a new TclObject with the
internal representation of \fIrep\fR.
.TP
\fBsetInternalRep\fR
The \fBsetInternalRep\fR method changes the internal representation
of the TclObject to the value \fIrep\fR.
.TP
\fBgetInternalRep\fR
The \fBgetInternalRep\fR method returns the internal representation
of the TclObject.  This methods allows you to query the class of a
returned object to determine what type of TclObject it is.
.TP
\fBtoString\fR
The \fBtoString\fR method returns the string representation of the
TclObject.
.TP
\fBinvalidateStringRep\fR
The \fBinvalidateStringRep\fR method marks the String representation
of the TclObject as invalid.  This method should be called only by
subclasses of the InternalRep prior to a call the
\fBsetInternalRep\fR.
.TP
\fBpreserve\fR
The \fBpreserve\fR method saves the value of a TclObject for
later use.  Each call to \fBpreserve\fR will internally increment the
"reference count" of the TclObject by one.
.TP
\fBrelease\fR
After saving the value of a TclObject for later use (via the
\fBpreserve\fR method), call the \fBrelease\fR method when you no
longer need the TclObject's value.  Each call to \fBrelease\fR will
internally decrement the "reference count" of the TclObject by one.
.TP
\fBisShared\fR
The \fBisShared\fR method returns true if the TclObject is shared (its
reference count is greater than one.)
.TP
\fBtakeExclusive\fR
The \fBtakeExclusive\fR method makes a copy of the TclObject if it is
shared.  The TclObject returned by the \fBtakeExclusive\fR will always
have a reference count of 1.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_NewObj, Tcl_ConvertToType, TclObj.typePtr, Tcl_GetStringFromObj,
Tcl_InvalidateStringRep, Tcl_IncrRefCount,
Tcl_DecrRefCount, Tcl_IsShared

.SH "SEE ALSO"
InternalRep, TclBoolean, TclDouble, TclList, TclIndex, TclInteger,
TclString, ReflectObject, Interp

.SH KEYWORDS
internal representation, object, object creation, object type,
reference counting, string representation, type conversion


ChangeLog
2019-05-04  Fabrice Ducos <fabrice.ducos@gmail.com>

        * Support of latest versions of Tcl (at least Tcl 8.5
        and Tcl 8.6) and of modern JDK (10+);
        support of JDK 4- abandoned (use of auto-boxing);
        support of OSX systems added;
        details below

        * configuration (tcljava.m4, configure.in),
        and compilation (Makefile.in) templates updated for
        JDK 10+ (javah not available anymore).
        javah is still used when it is available (JDK 9-).
        * minor update of jtclsh.in (for OSX dynamic libraries)
        * many obsolescent constructs (mostly old-style boxing)
        removed in *.java under src/jacl/tcl/lang,
        src/tclblend/tcl/lang and src/tcljava/tcl/lang,
        in order to silence many warnings.
        (effect: support of JDK 4- abandoned)
        * src/native/javaInterp.c: 2010 implementation
        of Java_tcl_lang_Interp_initName() crashed on the
        example given at docs/website/getstart.html,
        because calling Tcl_GetNameOfExecutable
        before Tcl_FindExecutable gives rise to undefined
        behaviour, at least in latest versions of Tcl
        (crashes observed on several systems, at least with
        Tcl 8.6).
        Java_tcl_lang_Interp_initName() is now simplified and
        always calls Tcl_FindExecutable first.
        * all the occurences of interp->result have been
        replaced with Tcl_GetStringResult(interp).
        Accessing interp->result directly was deprecated in
        latest versions of the Tcl C API, and removed from
        Tcl 8.6 (another solution was to #define USE_INTERP_RESULT,
        but this would have been an easy-going, quick and dirty way)
        ref. https://www.tcl.tk/man/tcl/TclLib/Interp.htm

2019-05-01  Fabrice Ducos <fabrice.ducos@gmail.com>

        * tcljava resurrected
        * migration from a CVS to a GIT repository
        (with reposurgeon)

2010-02-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* tests/tcl/regexp.test: Add test case for
        escaped - character in regexp command.
        [Bug 2012389]

2010-02-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceAll):
	Fixup regsub -all logic so 5 more tests pass.
	The replaceAll() logic is broken because it depends
	on matching substrings instead of using an index
	into the larger string and Java's regexp library
	uses this to calculate line starts in multiline mode.

2010-02-18  Mo DeJong  <mdejong@users.sourceforge.net>

        Incorporate fix/tests from Tcl head on 2010-02-11,
        aka was Tcl [Bug 2826551, Patch 2948425].

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
        Fixup increment index on zero length match issue.
	* tests/tcl/regexp.test: Back port of new regexp tests,
        some tests marked with {endIndex} due to lack of impl.

2010-02-11  Mo DeJong  <mdejong@users.sourceforge.net>

	Patch 2946118

	* src/jacl/tcl/lang/ListCmd.java: Minor style reformat.
	* src/jacl/tcl/lang/Util.java (scanElement, convertElement):
	Remove crusty old code that treated backslash zero
	as the end of the line. That logic was removed from
	C Tcl long ago.
	* tests/inprogress/EventAdaptor.test: Fixup tests
	that depended on backslash zero quirks.
	* tests/tcl/list.test: Add test cases for backslash zero
	as argument to list command.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (match, _match,
        replaceFirst, replaceAll): Rewrite -start offset
        logic to correctly deal with odd edge cases.
	* tests/tcl/regexp.test: Add additional test cases
        that cover -start index edge cases.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Rewrite argument processing logic to avoid using
        IndexOutOfBoundsException to deal with an incorrect
        number of arguments.

2009-10-03  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fix regsub -all crash because offset was not getting
        advanced by 1 on a match of length zero.
        Fixup -start argument support with and without -all
        in the case where the offset was larger than the
        length of the string.
	* tests/tcl/regexp.test: Add additional test cases.
        Enable regsub test cases that were getting caught
        in an infinite loop.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        failing tests related to optional subgroups.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Minor paren fix, no function changes.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Rework
        logic related to match result groups.
	* tests/tcl/regexp.test: Add missing test cases for
        non-capturing groups and -inline results.

2009-09-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup use of MULTILINE flag.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for -linestop and -lineanchor arguments to regexp.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):  Fixup
        support for -linestop and -lineanchor arguments to regsub.
	* tests/tcl/regexp.test: Add missing test cases for
        -linestop and -lineanchor arguments.

2009-09-22  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup support for the empty string as the string
        argument to regsub by removing the MULTILINE flag when
        an empty string is detected.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for empty string as string argument.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Remove comments.
	* tests/tcl/regexp.test: Add additional test cases

src/jacl/tcl/lang/TclObject.java
/*
 * TclObject.java
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclObject.java,v 1.6 2006/03/10 05:26:30 mdejong Exp $
 *
 */

package tcl.lang;

import java.util.Hashtable;

/**
 * This class extends TclObjectBase to implement the basic notion of
 * an object in Tcl.
 */

public final class TclObject extends TclObjectBase {

    static final boolean saveObjRecords = TclObjectBase.saveObjRecords;
    static Hashtable objRecordMap = TclObjectBase.objRecordMap;

    /**
     * Creates a TclObject with the given InternalRep. This method should be
     * called only by an InternalRep implementation.
     *
     * @param rep the initial InternalRep for this object.
     */
    public TclObject(final InternalRep rep) {
        super(rep);
    }

    /**
     * Creates a TclObject with the given InternalRep and stringRep.
     * This constructor is used by the TclString class only. No other place
     * should call this constructor.
     *
     * @param rep the initial InternalRep for this object.
     * @param s the initial string rep for this object.
     */
    protected TclObject(final TclString rep, final String s) {
        super(rep, s);
    }

    /**
     * Creates a TclObject with the given integer value.
     * This constructor is used by the TclInteger class only. No other place
     * should call this constructor.
     *
     * @param ivalue the integer value
     */
    protected TclObject(final int ivalue) {
        super(ivalue);
    }

    /**
     * Tcl_IncrRefCount -> preserve
     *
     * Increments the refCount to indicate the caller's intent to
     * preserve the value of this object. Each preserve() call must be matched
     * by a corresponding release() call. This method is Jacl specific
     * and is intended to be easily inlined in calling code.
     *
     * @exception TclRuntimeError if the object has already been deallocated.
     */
    public final void preserve() {
        if (refCount < 0) {
            throw DEALLOCATED;
        }
        refCount++;
    }

    /**
     * Tcl_DecrRefCount -> release
     *
     * Decrements the refCount to indicate that the caller is no longer
     * interested in the value of this object. If the refCount reaches 0,
     * the object will be deallocated. This method is Jacl specific
     * an is intended to be easily inlined in calling code.
     *
     * @exception TclRuntimeError if the object has already been deallocated.
     */
    public final void release() {
        if (--refCount <= 0) {
            disposeObject();
        }
    }

    /**
     * Return a String that describes TclObject and internal
     * rep type allocations and conversions. The string is
     * in lines separated by newlines. The saveObjRecords
     * needs to be set to true and Jacl recompiled for
     * this method to return a useful value.
     */

    public static String getObjRecords() {
        return TclObjectBase.getObjRecords();
    }
}

